### PR TITLE
EE-1093: Ensure gas injector works as intended.

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1093

This verifies that a) gas calls for batched opcodes can't be overflowed with
malicious WASM b) opcodes for wasm written by hand matches expectations.

### CODE REVIEW

File: grpc/tests/src/test/gas_counter.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    let opcode_costs = DEFAULT_WASM_CONFIG.opcode_costs();
+
+    const GROW_PAGES: u32 = 1;
+
+    let opcodes = vec![
+        (Some(opcode_costs.nop), Instruction::Nop),
+        (
+            Some(opcode_costs.current_memory),
+            Instruction::CurrentMemory(0),
+        ), // Push size to stack
```
Comment (Fraser999) [inline grpc/tests/src/test/gas_counter.rs:148]: I feel this test would benefit from some more comments explaining the rationale of the test itself, and in particular what this Vec represents.

File: grpc/tests/src/test/gas_counter.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            .with_authorization_keys(&[*DEFAULT_ACCOUNT_ADDR])
+            .with_deploy_hash([42; 32])
+            .build();
+        ExecuteRequestBuilder::from_deploy_item(deploy_item).build()
+    };
+
+    builder.exec(exec_request).commit().expect_success();
+
+    let gas_cost = builder.last_exec_gas_cost() - payment_cost;
+    let expected_cost = accounted_opcodes.clone().into_iter().map(Gas::from).sum();
```
Comment (Fraser999) [inline grpc/tests/src/test/gas_counter.rs:230]:
```suggestion
    let expected_cost = Gas::from(accounted_opcodes.iter().sum::<u32>());
```
If we do this, we can avoid adding `impl Sum for Gas` to gas.rs.

**Review by Fraser999 — APPROVED** [review]
Approved, but with a couple of non-blocking suggestions.

File: grpc/tests/src/test/gas_counter.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    let opcode_costs = DEFAULT_WASM_CONFIG.opcode_costs();
+
+    // Create a lot of `nop` opcodes to potentially overflow gas injector's batching counter.
+    let upper_bound = (u32::max_value() as usize / opcode_costs.nop as usize) + 1;
+
+    let instructions = {
+        let mut instructions = Vec::with_capacity(upper_bound);
+        for _ in 0..upper_bound {
+            instructions.push(Instruction::Nop);
+        }
```
[the remark this answers is not in the document]
Comment (mpapierski) [reply grpc/tests/src/test/gas_counter.rs:53]: Ah I knew it there's easier way 🤦‍♂️ Thanks!

File: grpc/tests/src/test/gas_counter.rs (hunk 2 above)
Comment (mpapierski) [reply grpc/tests/src/test/gas_counter.rs:227]: That's valid point though my intention here was to make a sum where partial values are Gas - similar to what EE would do during execution

File: grpc/tests/src/test/gas_counter.rs (hunk 1 above)
Comment (mpapierski) [reply grpc/tests/src/test/gas_counter.rs:145]: Did this

**Review by EdHastingsCasperLabs — APPROVED** [review]

**Pull request merged**